### PR TITLE
First commit of Phi In Jets task

### PR DESCRIPTION
**Pull request by smaff92**
First commit of a new analysis for Run 3 data, aimed to study resonances Phi and Kstar in/outside of jets

### CODE REVIEW

File: PWGJE/Tasks/PhiInJetsJE.cxx [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+
+/// \file PhiInJets.cxx
+/// \brief Reconstruction of Phi yield through track-track Minv correlations for resonance hadrochemistry analysis.
+///
+///
+/// \author Adrian Fereydon Nassirpour <adrian.fereydon.nassirpour@cern.ch>, Jimun Lee <jimun.lee@cern.ch>, Bong-Hwi Lim <bong-hwi.lim@cern.ch>, Sawan Sawan <sawan.sawan@cern.ch>
+
+#include <TLorentzVector.h>
+#include <TDatabasePDG.h> // FIXME
+#include <TPDGCode.h>     // FIXME
```
Comment (vkucera) [inline PWGJE/Tasks/PhiInJetsJE.cxx:20]: Please fix this

File: PWGJE/Tasks/PhiInJetsJE.cxx [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+
+    
+    //EVENT SELECTION
+    eventSelection = JetDerivedDataUtilities::initialiseEventSelection(static_cast<std::string>(c_eventSelections));
+    // trackSelection = JetDerivedDataUtilities::initialiseTrackSelection(static_cast<std::string>(trackSelections));
+
+  
+  } //end of init
+
+  double massKa = TDatabasePDG::Instance()->GetParticle(kKPlus)->Mass();  // FIXME: Get from the common header
```
Comment (vkucera) [inline PWGJE/Tasks/PhiInJetsJE.cxx:185]: Please fix this

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGJE/Tasks/PhiInJetsJE.cxx (hunk 1 above)
Comment (smaff92) [reply PWGJE/Tasks/PhiInJetsJE.cxx]: Dear Vit,

This has now been fixed.
Best Regards
Adrian.N

File: PWGJE/Tasks/PhiInJetsJE.cxx (hunk 2 above)
Comment (smaff92) [reply PWGJE/Tasks/PhiInJetsJE.cxx]: Dear Vit,

This has now been fixed.
Best Regards
Adrian.N

**Comment by vkucera** [comment]
Please do not force push during review.

File: PWGJE/Tasks/PhiInJetsJE.cxx [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+// or submit itself to any jurisdiction.
+
+/// \file PhiInJets.cxx
+/// \brief Reconstruction of Phi yield through track-track Minv correlations for resonance hadrochemistry analysis.
+///
+///
+/// \author Adrian Fereydon Nassirpour <adrian.fereydon.nassirpour@cern.ch>, Jimun Lee <jimun.lee@cern.ch>, Bong-Hwi Lim <bong-hwi.lim@cern.ch>, Sawan Sawan <sawan.sawan@cern.ch>
+
+#include <TLorentzVector.h>
+#include <TVector2.h>
```
Comment (vkucera) [inline PWGJE/Tasks/PhiInJetsJE.cxx:19]: You don't need these. Use `RecoDecay` features.

File: PWGJE/Tasks/PhiInJetsJE.cxx [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file PhiInJets.cxx
+/// \brief Reconstruction of Phi yield through track-track Minv correlations for resonance hadrochemistry analysis.
+///
+///
+/// \author Adrian Fereydon Nassirpour <adrian.fereydon.nassirpour@cern.ch>, Jimun Lee <jimun.lee@cern.ch>, Bong-Hwi Lim <bong-hwi.lim@cern.ch>, Sawan Sawan <sawan.sawan@cern.ch>
```
Comment (vkucera) [inline PWGJE/Tasks/PhiInJetsJE.cxx:16]: Put one author per line.

File: PWGJE/Tasks/PhiInJetsJE.cxx [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+#include "PWGJE/DataModel/Jet.h"
+
+#include "PWGLF/DataModel/LFResonanceTables.h"
+
+
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::framework::expressions;
+
+#include "Framework/runDataProcessing.h"
```
Comment (vkucera) [inline PWGJE/Tasks/PhiInJetsJE.cxx:51]: Move this up next to the other `Framework` includes.

File: PWGJE/Tasks/PhiInJetsJE.cxx [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+
+
+struct PhiInJetsJE {
+  SliceCache cache;
+  //Preslice<aod::Tracks> perCollision = aod::track::collisionId;
+  HistogramRegistry JEhistos{"JEhistos", {}, OutputObjHandlingPolicy::AnalysisObject};
+  HistogramRegistry LFhistos{"LFhistos", {}, OutputObjHandlingPolicy::AnalysisObject};
```
Comment (vkucera) [inline PWGJE/Tasks/PhiInJetsJE.cxx:58]: Why do you need two registries?

File: PWGJE/Tasks/PhiInJetsJE.cxx [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+
+
+  Configurable<std::string> c_eventSelections{"c_eventSelections", "sel8", "choose event selection"};
+  Configurable<std::string> c_trackSelections{"c_trackSelections", "globalTracks", "set track selections"};
```
Comment (vkucera) [inline PWGJE/Tasks/PhiInJetsJE.cxx:62]: Underscores are only allowed in names of namespaces.

File: PWGJE/Tasks/PhiInJetsJE.cxx [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+  using TrackCandidates = soa::Join<aod::Tracks, aod::TracksExtra, aod::TracksDCA, aod::TrackSelection,
+				    aod::pidTPCFullKa, aod::pidTOFFullKa>;  
+  Filter jetCuts = aod::jet::pt > c_jetPtMin&& aod::jet::r == nround(c_jetR.node() * 100.0f);
+
+
+  //Function for track quality cuts
+  template <typename TrackType>
+  bool trackSelection(const TrackType track){
+    // basic track cuts
+    if (std::abs(track.pt()) < c_trkMinPt)
```
Comment (vkucera) [inline PWGJE/Tasks/PhiInJetsJE.cxx:192]: You use `abs` where it is useless in many places.

File: PWGJE/Tasks/PhiInJetsJE.cxx [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+      return;
+
+    if (trk1.index() == trk2.index())
+      return; // We need to run (0,1), (1,0) pairs as well. but same id pairs are not needed.
+    
+    lDecayDaughter1.SetXYZM(trk1.px(), trk1.py(), trk1.pz(), massKa);
+    lDecayDaughter2.SetXYZM(trk2.px(), trk2.py(), trk2.pz(), massKa);
+    lResonance = lDecayDaughter1 + lDecayDaughter2;
+
+    if (abs(lResonance.Rapidity()) > 0.5)
```
Comment (vkucera) [inline PWGJE/Tasks/PhiInJetsJE.cxx:264]:
```suggestion
    if (std::abs(lResonance.Rapidity()) > 0.5)
```

File: PWGJE/Tasks/PhiInJetsJE.cxx [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+      
+    if(!trackPID(trk1) || !trackPID(trk2))
+      return;
+
+    if (trk1.index() == trk2.index())
+      return; // We need to run (0,1), (1,0) pairs as well. but same id pairs are not needed.
+    
+    lDecayDaughter1.SetXYZM(trk1.px(), trk1.py(), trk1.pz(), massKa);
+    lDecayDaughter2.SetXYZM(trk2.px(), trk2.py(), trk2.pz(), massKa);
+    lResonance = lDecayDaughter1 + lDecayDaughter2;
```
Comment (vkucera) [inline PWGJE/Tasks/PhiInJetsJE.cxx:262]: All this can be done with `RecoDecay`.

File: PWGJE/Tasks/PhiInJetsJE.cxx [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+
+      JEhistos.fill(HIST("hLSS"), mult, lResonance.Pt(), lResonance.M());	
+    }
+
+    bool jetFlag = false;
+    for (auto const& jet : jets){
+      
+      double phidiff = TVector2::Phi_mpi_pi(jet.phi()-lResonance.Phi());
+      double etadiff = jet.eta() - lResonance.Eta();     
+      double R = TMath::Sqrt((etadiff*etadiff) + (phidiff*phidiff));
```
Comment (vkucera) [inline PWGJE/Tasks/PhiInJetsJE.cxx:289]: All this can be done with `RecoDecay`.

File: PWGJE/Tasks/PhiInJetsJE.cxx [hunk 12]
```diff
@@ -0,0 +1,412 @@
...
+    
+    if (tpcPIDPassed && tofPIDPassed) {
+      return true;
+    }
+    return false;
+  }
+
+  
+  template <bool IsMC, bool IsMix, typename TracksType, typename JetType>
+  void MinvReconstruction(double mult, const TracksType& trk1, const TracksType& trk2, const JetType& jets) {
```
Comment (vkucera) [inline PWGJE/Tasks/PhiInJetsJE.cxx:248]: Names of functions and variables should start with small letters.

File: PWGJE/Tasks/CMakeLists.txt [hunk 13]
```diff
@@ -96,4 +96,8 @@ if(FastJet_FOUND)
                         SOURCES nSubjettiness.cxx
                         PUBLIC_LINK_LIBRARIES O2::Framework O2Physics::PWGJECore O2Physics::AnalysisCore
                         COMPONENT_NAME Analysis)
-endif()
\ No newline at end of file
+    o2physics_add_dpl_workflow(phi-in-jets
+                        SOURCES PhiInJetsJE.cxx
```
Comment (vkucera) [inline PWGJE/Tasks/CMakeLists.txt:100]:
```suggestion
                        SOURCES phiInJets.cxx
```
Workflow files should start with small letters. The suffix `JE` is redundant.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGJE/Tasks/PhiInJetsJE.cxx (hunk 4 above)
Comment (smaff92) [reply PWGJE/Tasks/PhiInJetsJE.cxx]: Fixed

File: PWGJE/Tasks/PhiInJetsJE.cxx (hunk 3 above)
Comment (smaff92) [reply PWGJE/Tasks/PhiInJetsJE.cxx]: I prefer to use these for now, as they are familiar to me. I will update them once I can confirm that my code works on hyperloop.

File: PWGJE/Tasks/PhiInJetsJE.cxx (hunk 5 above)
Comment (smaff92) [reply PWGJE/Tasks/PhiInJetsJE.cxx]: Fixed

File: PWGJE/Tasks/PhiInJetsJE.cxx (hunk 6 above)
Comment (smaff92) [reply PWGJE/Tasks/PhiInJetsJE.cxx]: Becasue I want one for LF tracks and one for JE tracks.

File: PWGJE/Tasks/PhiInJetsJE.cxx (hunk 7 above)
Comment (smaff92) [reply PWGJE/Tasks/PhiInJetsJE.cxx]: fixed

File: PWGJE/Tasks/PhiInJetsJE.cxx (hunk 10 above)
Comment (smaff92) [reply PWGJE/Tasks/PhiInJetsJE.cxx]: I prefer to use these for now, as they are familiar to me. I will update them once I can confirm that my code works on hyperloop.

File: PWGJE/Tasks/PhiInJetsJE.cxx (hunk 11 above)
Comment (smaff92) [reply PWGJE/Tasks/PhiInJetsJE.cxx]: I prefer to use these for now, as they are familiar to me. I will update them once I can confirm that my code works on hyperloop.

File: PWGJE/Tasks/PhiInJetsJE.cxx (hunk 12 above)
Comment (smaff92) [reply PWGJE/Tasks/PhiInJetsJE.cxx]: fixed

File: PWGJE/Tasks/CMakeLists.txt (hunk 13 above)
Comment (smaff92) [reply PWGJE/Tasks/CMakeLists.txt]: fixed

**Comment by smaff92** [comment]
Dear Vit,

Unless you find any major errors in the code, could you please just approve the merge?
Then I will implement any minor suggestions for the next iteration of the code, for now I just want to test that I can run it on hyperloop.

Best Regards
Adrian.N

File: PWGJE/Tasks/PhiInJetsJE.cxx (hunk 3 above)
Comment (vkucera) [reply PWGJE/Tasks/PhiInJetsJE.cxx]: OK. But please come back to it. `TLorentzVector` is unnecessarily heavy here.

File: PWGJE/Tasks/PhiInJetsJE.cxx (hunk 4 above)
Comment (vkucera) [reply PWGJE/Tasks/PhiInJetsJE.cxx]: Why did you remove the other names? Are they not authors?

File: PWGJE/Tasks/PhiInJetsJE.cxx (hunk 6 above)
Comment (vkucera) [reply PWGJE/Tasks/PhiInJetsJE.cxx]: But what difference does it make? All the histograms end up in the same folder anyway, no?

**Review by vkucera — APPROVED** [review]
Hi @smaff92 , thanks for addressing some of the comments. Please fix the remaining issues once you get your code running on Hyperloop.
You still need to fix the formatting by merging https://github.com/smaff92/O2Physics/pull/1 and approval from the code owners before your PR can be merged.

**Comment by ddobrigk** [comment]
@nzardosh @maoyx @mfasDa shall this be approved? Vit has agreed to merge as is pending future work. Please let me know - if no feedback arrives by the end of today I will approve from my side. Thanks!

**Review by ddobrigk — APPROVED** [review]
Approving since no reaction from approvers and reviewer is satisfied for now. Thanks!

**Comment by smaff92** [comment]
It seems that PWGHF/Core/PDG.h got removed two days ago?
I moved over to that header file upon Vit's first request.
For the sake of stability, I will now move back to <TDatabasePDG.h> and <TPDGCode.h>.

During the review process I have made major additions to the code, it now includes MC process functions as well (although the code data code has remained unchanged.)

**Comment by vkucera** [comment]
> It seems that PWGHF/Core/PDG.h got removed two days ago? I moved over to that header file upon Vit's first request. For the sake of stability, I will now move back to <TDatabasePDG.h> and <TPDGCode.h>.
> 
> During the review process I have made major additions to the code, it now includes MC process functions as well (although the code data code has remained unchanged.)

`PWGHF/Core/PDG.h` was replaced with [PhysicsConstants.h](https://github.com/AliceO2Group/AliceO2/blob/dev/Common/Constants/include/CommonConstants/PhysicsConstants.h). Please check the changes in https://github.com/AliceO2Group/O2Physics/pull/3976 for examples of how to modify your code accordingly. Do not revert back to the ROOT headers. Please follow the O2 announcement channel to stay up to date with the news.

**Comment by smaff92** [comment]
Dear @vkucera 

I have now changed the code to utilize PhysicsConstants.h
It compiles locally after a fresh git pull.

Could you please proceed with the test build?

Best Regards
Adrian.N

**Comment by vkucera** [comment]
Dear @smaff92 , thanks for the update.
The build tests run automatically. If this is your first PR in O2Physics, you will need an approval from the code owners to enable them. (Check the [contribution guidelines in the documentation](https://aliceo2group.github.io/analysis-framework/docs/gettingstarted/contributingtocode.html).)
In the meantime you have to fix the formatting to make your PR ready for merging.

**Comment by smaff92** [comment]
Dear @nzardosh, @maoyx, @mfasDa 

I just wanted to ping a request for approving the merge of my code into O2Physics.
If done today, I might be able to have results ready for the next PAG-JE meeting.
Thanks!

Best Regards
Adrian.N

**Review by nzardosh — APPROVED** [review]

**Pull request merged**